resources/benchmark-runner.mjs

```diff
@@ -116,6 +116,106 @@ class PageElement {
     }
 }
 
+class MeasureTask {
+    constructor(runner, suite, test, page, frame) {
+        this._runner = runner;
+        this.suite = suite;
+        this.test = test;
+        this._page = page;
+        this._frame = frame;
+
+        this._unusedHeight = undefined;
+
+        this._wasRun = false;
+
+        this._asyncMeasurePromise = new Promise((resolve) => {
+            this._asyncDoneCallback = resolve;
+        });
+        this._measureSyncCallback = this._measureSync.bind(this);
+        this._measureAsyncTimeoutCallback = this._measureAsyncTimeout.bind(this);
+        this._measureAsyncRafCallback = this._measureAsyncRaf.bind(this);
+
+        this.syncTime = 0;
+        this.asyncTime = 0;
+        this.layoutTime = 0;
+
+        this._asyncStartTime = 0;
+
+        const label = `${suite.name}.${test.name}`;
+        this._startLabel = `${label}-start`;
+        this._syncEndLabel = `${label}-sync-end`;
+        this._asyncStartLabel = `${label}-async-start`;
+        this._asyncEndLabel = `${label}-async-end`;
+        this._forcedLayoutStartLabel = `${label}-layout-start`;
+        this._forcedLayoutEndLabel = `${label}-layout-end`;
+    }
+
+    async run() {
+        if (this._wasRun)
+            throw Error("MeasureTask can only run once.");
+        if (params.asyncMetric === "raf") {
+            requestAnimationFrame(this._measureSyncCallback);
+            requestAnimationFrame(this._measureAsyncRafCallback);
+        } else {
+            setTimeout(this._measureSyncCallback, 0);
+        }
+        await this._done();
+        this._wasRun = true;
+    }
+
+    _measureSync() {
+        // This function ought be as simple as possible. Don't even use Promise.
+        performance.mark(this._startLabel);
+        const syncStartTime = performance.now();
+        this.test.run(this._page);
+        const syncEndTime = performance.now();
+        performance.mark(this._syncEndLabel);
+
+        this.syncTime = syncEndTime - syncStartTime;
+
+        performance.mark(this._asyncStartLabel);
+        this._asyncStartTime = performance.now();
+        if (params.asyncMetric === "timeout")
+            setTimeout(this._measureAsyncTimeoutCallback, 0);
+    }
+
+    _measureForceLayout() {
+        performance.mark(this._forcedLayoutStartLabel);
+        const startTime = performance.now();
+        this._unusedHeight = this._frame.contentDocument.body.getBoundingClientRect().height;
+        const endTime = performance.now();
+        performance.mark(this._forcedLayoutEndLabel);
+        this.layoutTime = endTime - startTime;
+        // Prevent dead code elimination.
+        this._frame.contentWindow._unusedHeightValue = this._unusedHeight;
+    }
+
+    _measureAsyncTimeout() {
```

Comment:
Since we're using rAF, we should probably call this _measureAsyncEnd or something.